parse/sql/sqlanalyzer/typing/visitor.go

```diff
@@ -562,260 +486,241 @@ func (t *typeVisitor) VisitExpressionColumn(p0 *tree.ExpressionColumn) any {
 		if p0.TypeCast != nil {
-			return anonAttr(p0.TypeCast), nil
+			return anonAttr(p0.TypeCast)
 		}
 
-		return anonAttr(types.NullType), nil
+		return anonAttr(types.NullType)
 	})
 }
 
 func (t *typeVisitor) VisitExpressionBlobLiteral(p0 *tree.ExpressionBlobLiteral) any {
-	return attributeFn(func(ev *evaluationContext) (*QualifiedAttribute, error) {
+	return attributeFn(func(ev *evaluationContext) *QualifiedAttribute {
 		if p0.TypeCast != nil {
-			return anonAttr(p0.TypeCast), nil
+			return anonAttr(p0.TypeCast)
 		}
 
-		return anonAttr(types.BlobType), nil
+		return anonAttr(types.BlobType)
 	})
 }
 
 func (t *typeVisitor) VisitExpressionSelect(p0 *tree.ExpressionSelect) any {
-	return attributeFn(func(ev *evaluationContext) (*QualifiedAttribute, error) {
-		r, err := p0.Select.Accept(t).(returnFunc)(ev)
-		if err != nil {
-			return nil, err
-		}
+	return attributeFn(func(ev *evaluationContext) *QualifiedAttribute {
+		r := p0.Select.Accept(t).(returnFunc)(ev)
 
 		shape := r.Shape()
 		if len(shape) != 1 && !p0.IsExists {
-			return nil, fmt.Errorf("subquery must return exactly one column")
+			t.options.ErrorListener.NodeErr(p0.GetNode(), parseTypes.ParseErrorTypeSemantic, errors.New("subquery must return exactly one column"))
+			return unknownAttr()
 		}
 
 		if p0.TypeCast != nil {
-			return anonAttr(p0.TypeCast), nil
+			return anonAttr(p0.TypeCast)
 		}
 
 		if p0.IsExists {
-			return anonAttr(types.BoolType), nil
+			return anonAttr(types.BoolType)
 		}
 
-		return anonAttr(shape[0]), nil
+		return anonAttr(shape[0])
 	})
 }
 
 func (t *typeVisitor) VisitExpressionStringCompare(p0 *tree.ExpressionStringCompare) any {
-	return attributeFn(func(ev *evaluationContext) (*QualifiedAttribute, error) {
+	return attributeFn(func(ev *evaluationContext) *QualifiedAttribute {
 		a := p0.Left.Accept(t).(attributeFn)
 		b := p0.Right.Accept(t).(attributeFn)
 
 		// do these both need to be text? I believe so
-		at, err := a(ev)
-		if err != nil {
-			return nil, err
-		}
-		bt, err := b(ev)
-		if err != nil {
-			return nil, err
-		}
+		at := a(ev)
+		bt := b(ev)
+
 		if !at.Type.Equals(bt.Type) {
-			return nil, fmt.Errorf("%w: string comparison expression expected %s. Received: %s", ErrInvalidType, at.Type.String(), bt.Type.String())
+			t.options.ErrorListener.NodeErr(p0.GetNode(), parseTypes.ParseErrorTypeType, fmt.Errorf("string comparison expression expected %s. Received: %s", at.Type.String(), bt.Type.String()))
+			return unknownAttr()
 		}
 
 		if p0.Escape != nil {
 			esc := p0.Escape.Accept(t).(attributeFn)
-			et, err := esc(ev)
-			if err != nil {
-				return nil, err
-			}
+			et := esc(ev)
 
 			if !et.Type.Equals(types.TextType) {
-				return nil, fmt.Errorf("%w: string comparison expected text. Received: %s", ErrInvalidType, et.Type.String())
+				t.options.ErrorListener.NodeErr(p0.GetNode(), parseTypes.ParseErrorTypeType, fmt.Errorf("string comparison escape expected text. Received: %s", et.Type.String()))
```

Comment:
escape?

Reply:
yeah, SQL string comparisons allow escape clauses for escaping variables from the comparison